lib/src/model.dart

```diff
@@ -2430,19 +2428,30 @@ class Library extends ModelElement with Categorization, TopLevelContainer {
   CompilationUnitElement get compilationUnitElement => (element as LibraryElement).definingCompilationUnit;
 
   @override
-  Iterable<Class> get classes {
-    return _allClasses
-        .where((c) => !c.isErrorOrException)
-        .toList(growable: false);
-  }
+  Iterable<Class> get classes => allClasses.where((c) => !c.isErrorOrException);
 
   @override
   Iterable<Extension> get extensions {
-    if (_extensions != null) return _extensions;
-    _extensions = _libraryElement.definingCompilationUnit.extensions
-        .map((e) => ModelElement.from(e, this, packageGraph) as Extension)
-        .toList(growable: false)
-          ..sort(byName);
+    if (_extensions == null) {
+      // De-dupe extensions coming from multiple exported libraries at once.
+      Set<ExtensionElement> extensionElements = Set();
+      extensionElements.addAll(_libraryElement.definingCompilationUnit.extensions);
+      for (CompilationUnitElement cu in _libraryElement.parts) {
+        extensionElements.addAll(cu.extensions);
+      }
+      for (LibraryElement le in _libraryElement.exportedLibraries) {
+        extensionElements.addAll(le.definingCompilationUnit.extensions
+            .where((t) => _exportedNamespace.definedNames.values.contains(t.name)));
```

Comment:
I'm worried this might not do the right thing.  Consider files `foo.dart` and `bar.dart`, each of which defines a distinct extension called `E`.  Then if we have a library like this:
```dart
export 'foo.dart' show E;
export 'bar.dart' hide E;
```
Then I think both `E`s will get included in `extensionElements`, because both of them have a name that matches a name that exists in `_exportedNamespace`.

The best correct approach I can currently think of is:
- Expose an analyzer API to allow clients to figure out which elements are actually exported by an `ExportElement` (e.g. we could expose  dartdoc to filter a set of elements via namespace combinators (we could move `NamespaceBuilder._applyCombinators` to `ExportElement` and re-name it appropriately)
- Change this code to iterate through the export elements rather than the exported libraries, so that it can use the newly exposed API to figure out exactly which extensions are shown.

It's ok with me if you want to file an issue and address this in a later PR.

Reply:
Yes, I'm seeing this now that I'm looking more closely at the library scope code in the analyzer.  It looks like this has always been wrong in the class implementation.  The "LibraryExportScope" I mentioned in chat the other day is probably the right solution for this, too.

Reply:
The fix for this is taking a while, I think I will go ahead and deal with this in a followup PR.